File: solution/0001-0100/0002-add-two-numbers/solution.py
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummyNode = ListNode()
        carry = 0
        cur = dummyNode
        while l1 != None or l2 != None or carry != 0:
            sum = (0 if l1 == None else l1.val) + (0 if l2 == None else l2.val) + carry
            carry, val = divmod(sum, 10)
            cur.next = ListNode(val=val)
            cur = cur.next
            l1 = None if not l1 else l1.next
            l2 = None if not l2 else l2.next
        return dummyNode.next

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

File: solution/0001-0100/0002-add-two-numbers/Solution.java
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummyNode = new ListNode(0);
        int carry = 0;
        ListNode p = dummyNode;
        while(l1 != null || l2 != null || carry != 0){
            if(l1 != null){
                carry += l1.val;
                l1 = l1.next;
            }
            if(l2 != null){
                carry += l2.val;
                l2 = l2.next;
            }
            p.next = new ListNode(carry % 10);
            p = p.next;
            carry /= 10;
        }
        return dummyNode.next;
    }

    public ListNode addTwoNumbers2(ListNode l1, ListNode l2){
        if(l1 == null){
            return l2;
        }
        if(l2 == null){
            return l1;
        }
        ListNode dummyNode = new ListNode(0);
        ListNode p = dummyNode;
        int carry = 0;
        while(l1 != null && l2 != null){
            int sum = l1.val + l2.val + carry;
            p.next = new ListNode(sum % 10);
            p = p.next;
            l1 = l1.next;
            l2 = l2.next;
            carry = sum / 10;
        }
        while(l1 != null){
            int sum = l1.val + carry;
            p.next = new ListNode(sum % 10);
            p = p.next;
            l1 = l1.next;
            carry = sum / 10;
        }
        while(l2 != null){
            int sum = l2.val + carry;
            p.next = new ListNode(sum % 10);
            p = p.next;
            l2 = l2.next;
            carry = sum / 10;
        }
        if(l1 == null && l2 == null && carry > 0){
            p.next = new ListNode(carry);
            p = p.next;
        }
        return dummyNode.next;
    }
}

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}